service/src/main/java/edu/tamu/catalog/service/FolioCatalogService.java

```diff
@@ -367,6 +399,78 @@ private String getServicePointDisplayName(String servicePointId) {
         return  null;
     }
 
+    /**
+     * Get FOLIO User by the user's UIN via OKAPI.
+     *
+     * @param uin The user's UIN.
+     * @return request type
+     */
+    private JsonNode getUserByUin(String uin) {
+        if (Objects.isNull(uin)) {
+            return null;
+        }
+
+        String path = "%s/bl-users?query=(externalSystemId==\"{uin}\")&limit=2";
+        String url = String.format(path, properties.getBaseOkapiUrl(), uin);
+        String message = "user using external system id";
+
+        logger.debug("Asking for User from: {}", url);
+
+        JsonNode response = okapiRequestJsonNode(url, HttpMethod.GET, message, uin);
+
+        if (Objects.nonNull(response) && response.isObject()) {
```

Comment:
JsonNode is never null. Use isMissionNode. https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/JsonNode.html#isMissingNode()

Reply:
I think the function `isMissingNode()` is not what you think it is in this case:

I think "missingNode" is not "a node is missing" but instead is a type of node called a "missing node".
[At least that is how I am reading the documentation.](http://fasterxml.github.io/jackson-core/javadoc/2.7/com/fasterxml/jackson/core/TreeNode.html?is-external=true#isMissingNode())

We are also using Spring-Framework 4.3.x.
[The code eventually calls doExecute().](https://github.com/spring-projects/spring-framework/blob/21648426e942e97292c60aa104d809d9dab1ea9b/spring-web/src/main/java/org/springframework/web/client/RestTemplate.java#L650)

[What happens within this function is that it does potentially return NULL.](https://github.com/spring-projects/spring-framework/blob/21648426e942e97292c60aa104d809d9dab1ea9b/spring-web/src/main/java/org/springframework/web/client/RestTemplate.java#L667)

Reply:
That is not what I think it is. A JsonNode is always one of the three, missing node, container node, or value node. Never null.

Reply:
http://fasterxml.github.io/jackson-core/javadoc/2.7/com/fasterxml/jackson/core/TreeNode.html?is-external=true#isMissingNode()

> Note: one and only one of methods isValueNode(), isContainerNode() and isMissingNode() ever returns true for any given node.

Reply:
We primarily use exchange or other RestTemplate model that is not annotated with @Nullable. https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/client/RestTemplate.html#exchange-org.springframework.http.RequestEntity-java.lang.Class-

We do not use public execute method which is annotated Nullable, https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/client/RestTemplate.html#execute-java.lang.String-org.springframework.http.HttpMethod-org.springframework.web.client.RequestCallback-org.springframework.web.client.ResponseExtractor-java.util.Map-

Reply:
The function `okapiRequestJsonNode()` ultimately returns a call from `restTemplate.exchange()` which is what returns the JsonNode.
[See this link.](https://github.com/TAMULib/CatalogService/blob/fb16d6d79d9da51e1a72246ca4a9dedf09aec834/service/src/main/java/edu/tamu/catalog/service/FolioCatalogService.java#L744)

The function `restTemplate.exchange()` eventually calls `restTemplate.doExecute()`, which does return NULL.
As I referenced to in the links in the above comment.

And as for:
```
Note: one and only one of methods isValueNode(), isContainerNode() and isMissingNode() ever returns true for any given node.
```
I know, I referenced that exact documentation in a link in my previous comment and described how it is more like "a type" and less like a "this does not exist".
Using different words, we are potentially getting a NULL response.
`NULL.isMissingNode()` is an error.
It doesn't matter what methods are on the JsonNode model, if we have NULL there is no model to operate on.

Reply:
And what I am saying we should never return null for a JsonNode and not use any methods that can.

Reply:
Just because the protected doExecute method can return null does not mean the exchange method does. In fact there documentation @Nullable indicates doExecute can and it is not on exchange. Which tells me exchange will never return null.

Reply:
And what I am saying is that it is **Spring**, not us, returning that NULL.

If you want to throw an Internal Server Error **after spring returns NULL**, then we can but we still have to check for that NULL and handle it.
The current behavior is consistent with handling what Spring RestTemplate is returning to us.

Reply:
How are you so sure RestTemplate will ever throw null? I don't think it does.

Reply:
https://github.com/TAMULib/CatalogService/blob/fb16d6d79d9da51e1a72246ca4a9dedf09aec834/service/src/main/java/edu/tamu/catalog/service/FolioCatalogService.java#L742

Spring returns a ResponseEntity and the body is the JsonNode. I just don't see where it will ever become null. In any case if the body is null we should only do a check in one location to avoid repeating logic everywhere.

Reply:
Also, those null checks of the JsonNode would be if getBody returned null.

Reply:
I believe "leverage exception handling" is a separate argument.

The current code does leverage exception handling.
It catches and treats a NULL response as a 404 Not Found.
Not catching a NULL response and letting it explode would result in a 500 Internal Server Error.

This is an argument not of whether or not we are getting and checking for NULL.
Instead, it is an argument on whether we want to return 404 or 500 when we do get a NULL.
Furthermore, instead of letting in explode on NULL, we handle the NULL before it does and provide a more detailed exception message because this is a specific NULL pointer. This gives us granularity on the error.

There is also room for discussion on instead returning a 400 Bad Request.
It could be that the NULL is only returned if bad data is sent, so 500 and 404 both may be less correct than a 400.

Reply:
A NullPointerException can be handled the same... globally. I do not see how that is a separate argument. I just don't want a bunch of repeated conditions that never diverges to false. A way you can prove it can be null, is to write a reasonable test that shows the divergence.

Reply:
We are determining the more precise problem and reporting it to the client so that client knows that, in this case, it failed to find the user.
Getting the more precise response is more useful when utilizing the CatalogService API.
`404: Unable to find user.` is more friendly and more accurate, than `500: Internal Server Error"`
This is further better practice as it gives easier to diagnose and debug making it more likely that 3rd-parties adapt our APIs.

(Also, I believe showing the actual code literally returning NULL (as I have done so in the previous comments) is pretty sufficient proof. It is not a possibility, it is a fact.)

Reply:
Show me the coverage of the case where those are false then. And I am not trying to change the fact of return meaningful response messages on failure.

Reply:
https://coveralls.io/builds/38143009/source?filename=service%2Fsrc%2Fmain%2Fjava%2Fedu%2Ftamu%2Fcatalog%2Fservice%2FFolioCatalogService.java

Well darn cant see the coverage on these.

Reply:
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/HttpEntity.html#getBody--

Looks like ResponseEntity getBody is Nullable. We should perform the null check in the method that calls getBody.

Reply:
I do not think simple what you showed of doExecute returning a null is sufficient evidence that it would result in the method calls we are making will return null. The only condition in which it returns null is if responseExtractor is null.

https://github.com/spring-projects/spring-framework/blob/21648426e942e97292c60aa104d809d9dab1ea9b/spring-web/src/main/java/org/springframework/web/client/RestTemplate.java#L662

As responseExtractor is an argument we have to look further in method calls that we do use and see how they pass responseExtractor to doExecute. It looks as all exchange methods use the following function to instantiate the responseExecutor and that will never be null thus doExecute will never return null for calls to exchange that we use.

https://github.com/spring-projects/spring-framework/blob/21648426e942e97292c60aa104d809d9dab1ea9b/spring-web/src/main/java/org/springframework/web/client/RestTemplate.java#L741


This is all for naught as it is actually the getBody method from ResponseEntity that will be returning the typed response body JsonNode. The getBody method is inherited from HttpEntity which has a convenience method to perform the null check called hasBody. Whether this will ever be null from a ResponseEntity returned from RestTemplate is also still not certain. It may be that somewhere in the deserialization phase of fasterxml integration the body will end up being null. But as it goes it will likely follow a better practice of throwing an exception rather than having a consumer of an API perform excessive null checks on its responses.